apply/apply.go

```diff
@@ -52,24 +62,52 @@ func Apply(fs afero.Fs, conf *config.Config, tmp *templates.T, siccMode bool) er
 	}
 
 	e = applyModules(fs, p.Modules, &tmp.Module)
-
 	return errors.Wrap(e, "unable to apply modules")
 }
 
 func applyRepo(fs afero.Fs, p *plan.Plan, repoTemplates *packr.Box) error {
-	e := applyTree(repoTemplates, fs, "", p.SiccMode, p)
+	e := applyTree(fs, repoTemplates, "", p.SiccMode, p)
 	if e != nil {
 		return e
 	}
 	// Remove after migration.
-
 	if !p.SiccMode {
 		e := fs.Remove(".sicc-version")
 		if e != nil {
 			log.Debug("error removing .sicc-version. ignoring")
 		}
 	}
-	return nil
+	return applyCustomProviders(fs, p)
+}
+
+func applyCustomProviders(fs afero.Fs, p *plan.Plan) (err error) {
+	for providerName, customProvider := range p.CustomProviders {
+		log.Infof("Applying custom provider %s", providerName)
+		err = customProvider.Install(fs, providerName)
+		if err != nil {
+			return errors.Wrapf(err, "Error applying custom provider %s", providerName)
+		}
+	}
+	return
+}
+
+func applyCustomPluginCache(fs afero.Fs, p *plan.Plan, repoBox *packr.Box) error {
+	path := providers.CustomPluginCacheDir
+	err := fs.MkdirAll(path, 0755)
+	if err != nil {
+		return errors.Wrapf(err, "unable to make directory %s", path)
+	}
+	return applyTree(fs, repoBox, path, p.SiccMode, p)
+}
+
+// Terraform won't cache providers unless the directory structure already exists
+func applyPluginCache(fs afero.Fs, p *plan.Plan, repoBox *packr.Box) error {
+	path := providers.PluginCacheDir
+	err := fs.MkdirAll(path, 0755)
+	if err != nil {
+		return errors.Wrapf(err, "unable to make directory %s", path)
+	}
+	return applyTree(fs, repoBox, path, p.SiccMode, p)
```

Comment:
and here?

Reply:
note that if we just put a file in those directories in templates/repo it will do it for us. maybe .gitignore?